design-proposals/klip-57-kafka-headers.md

```diff
@@ -0,0 +1,140 @@
+# KLIP 57 - Reading Kafka headers in ksqlDB
+**Author**: Zara Lim (@jzaralim) | 
+**Release Target**: 0.25 | 
+**Status**: _In Discussion_ | 
+**Discussion**: https://github.com/confluentinc/ksql/pull/8293
+
+**tl;dr:** _Expose Kafka header data in ksqlDB. Currently, headers are the only metadata in a Kafka record that is invisible to users._
+
+## Motivation and background
+
+So far, ksqlDB exposes the timestamp, partition and offset of Kafka records as pseudocolumns. The only metadata left that is invisible to ksqlDB users are [headers](https://kafka.apache.org/28/javadoc/org/apache/kafka/common/header/Headers.html). Kafka headers are a list of zero or more key-value pairs, where the keys are (not necessarily unique) strings and the values are byte arrays. Headers typically contain metadata about records, which can then be used for routing or processing (e.g. the header could store information about how to deserialize strings in the value). If ksqlDB exposes headers to the user, then these steps could be done in ksqlDB.
+
+## What is in scope
+* Syntax for accessing headers
+* Support for all query types except for `INSERT` queries
+* Including header columns in a data source's schema
+* Functions to help decode byte strings to primitive types (BYTES to STRING already exists, new functions would be conversions to INT, BIGINT and DOUBLE)
+
+## What is not in scope
+* Writing headers to output topics - we might do this in the future, but that is a separate discussion
+* Advanced functions to help decode byte strings (e.g. functions to deserialize complex serialized data.) - if a user wants to convert something that we don't have a function for, then they can either create their own UDF or use `CREATE FUNCTION` when that becomes available.
+
+## Public APIS
+
+### Syntax
+
+Unlike other pseudocolumns, if a user wants to access headers, they will create a column with the `HEADERS` keyword. This column will be populated with the full list of header keys and values. There can be any number of HEADERS columns.
+```
+ksql> CREATE STREAM A (
+    my_headers ARRAY<STRUCT<key STRING, value BYTES>> HEADERS,
+    my_headers_2 ARRAY<STRUCT<key STRING, value BYTES>> HEADERS);
+```
```

Comment:
I think we might have issues with multiple HEADERS section in the future when we support inserts. For instance, trying to insert the same keys with different values in both headers. Which one would be the last key written? I think we should allow only one HEADER section. @colinhicks what do you think?

Reply:
For the write path, I agree that we need to limit to a single header. But it seems orthogonal to allow `HEADERS` multiple times to define an "input" schema.

For the "input" schema, you would get multiple columns reading the same headers (which is redundant and not useful) but also not harmful.

For the write path, we don't have a way to write into the header atm, and if a header is selected right now, it would be copied into the record _value_ (so no issue so far). My proposal would be to add a new function `AS_HEADER()` (similar to `AS_VALUE()` that copies a key column into the record value), and we can limit that `AS_HEADER()` is used at max once in the `SELECT` clause.

Reply:
Yeah, I support what @mjsax is suggesting. I think for insert-into-values, too, users would need to use a function like `AS_HEADER()`, rather than assigning the headers based on the logical schema's column name. Maybe we would even provide an `AS_HEADER(<header-key>, <header-value>)` signature to make this more convenient ...

```
INSERT INTO A () VALUES (AS_HEADER('my-header-key', 'my-header-value'));
```

We can hash out the details when the time comes to support writes.

Reply:
Yes, multiple HEADERS columns is not harmful, but I think we should not allow multiple headers yet if we haven't hash out the details for the insert statements. The idea of the AS_HEADER sounds good, but what if there are other recommendations later? others that will conflict with the multiple headers and we cannot use them because users already using multiple headers? I prefer to not support that until we have a good design for how we'll insert headers.

Reply:
`INSERT INTO` is actually an excellent point. So far, for the write path, we only discussed persistent queries (CSAS/CTAS). However, particularly `INSERT INTO VALUES` is a different write path that we did not discuss yet. (Not 100% if `INSERT INTO AS SELECT` would fall into the CSAS/CTAS bucket or the `INSERT INTO VALUES` bucket?)

I guess we could actually support `INSERT INTO VALUES` directly in the first version: in the end, the type of the header column is well defined, and thus, if a user provides as with a valid `STRUCT`, there is no reason not to write it into the header field of the record. It's of course more work for (v1) but it might not be too complex? If we don't want to support it, the question is if a stream/table with a header column would need to reject _all_ `INSERT INTO VALUES` statements? Or could we require the that header column is `NULL` as long as we don't support `INSERT INTOE VALUES`?

In any case, it would be weird to have two header columns for `INSERT INTO VALUES`, because it would not be well defined how we "merge" both into a single record header field. As allowing multiple HEADER columns does not really buy us anything, but disallow multiple avoid the issue for `INSERT INTO VALUES`, I am convinced now, that we should strictly enforce a single HEADER column in a schema definition.

Reply:
That makes sense. So to recap:

1. We'll limit to one `HEADERS` column.
2. If there is a `HEADERS` column, then we cannot have any `HEADER(key)` columns.
3. There can still be multiple `HEADER(key)` columns, since merging them all isn't as confusing.

From an implementation standpoint, `INSERT INTO VALUES` might not be too hard, and i would testing more convenient. I think it should be a stretch goal though, since this klip only promises reading kafka headers.

Reply:
Should prevent multiple `HEADER(key)` that uses the same `key` too?

Reply:
I was thinking that since it's possible for a header to repeat keys, merging won't be a problem if we can decide on how to order them. Though now that I think of it, this is basically the same problem as before, but on a smaller scale, so it makes sense to prevent the same `key` from being used multiple times. I'll update the KLIP with the restrictions.